<a href="https://colab.research.google.com/github/darsh0820/courses/blob/main/Machine%20Learning/ML_Basics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Introduction to Machine Learning
Machine Learning is a program that analyses data and learns to predict the outcome.

##Data Set
In the mind of a computer, a data set is any collection of data. It can be anything from an array to a complete database.

Example of an array:
```
[99,86,87,88,111,86,103,87,94,78,77,85,86]
```
By looking at the array, we can guess that the average value is probably around 90, and we are also able to determine that the highest value is 111 and the lowest value is 77.

Suppose we have a database consisting of data about house pricing based on area, number of rooms, number of bedrooms etc. We can use machine learning to predict the price of a home based on those provided factors.

In Machine Learning it is common to work with very large data sets.

##Data Types
To analyze data, it is important to know what type of data we are dealing with.

We can split the data types into three main categories:

- **Numerical data** are numbers, and can be split into two numerical categories:

  Discrete Data
  - counted data that are limited to integers. Example: A person's age.
  Continuous Data
  - measured data that can be any number. Example: The price of an item.

- **Categorical data** are values that cannot be measured up against each other. Example: a color value, or any yes/no values.

- **Ordinal data** are like categorical data, but can be measured up against each other. Example: school grades where A is better than B and so on.

#Mean, Median, and Mode

In Machine Learning there are often three mathematical values that interests us:
- Mean
- Median
- Mode

Example: We have registered the speed of 13 cars:
```
speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]
```

## Mean
The mean value is the average value.

Use the NumPy `mean()` method to find the average speed:

In [1]:
import numpy

speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

x = numpy.mean(speed)

print(x)

89.76923076923077


## Median
The median value is the value in the middle, after all the values have been sorted.

Use the NumPy `median()` method to find the middle value:

In [2]:
import numpy

speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

x = numpy.median(speed)

print(x)

87.0


If there are two numbers in the middle, divide the sum of those numbers by two.
```
77, 78, 85, 86, 86, 86, 87, 87, 94, 98, 99, 103
(86 + 87) / 2 = 86.5
```

In [3]:
import numpy

speed = [99,86,87,88,86,103,87,94,78,77,85,86]

x = numpy.median(speed)

print(x)

86.5


##Mode
The Mode value is the value that appears the most number of times.

Use the SciPy `mode()` method to find the number that appears the most:

In [4]:
from scipy import stats

speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

x = stats.mode(speed)

print(x)

ModeResult(mode=86, count=3)


#Standard Deviation
Standard deviation is a number that describes how spread out the values are.
A low standard deviation means that most of the numbers are close to the mean value whereas a high standard deviation means that the values are spread out over a wider range.

Use the NumPy std() method to find the standard deviation:


In [5]:
import numpy

speed = [86,87,88,86,87,85,86]

x = numpy.std(speed)

print(x)

0.9035079029052513


In [6]:
import numpy

speed = [32,111,138,28,59,77,97]

x = numpy.std(speed)

print(x)

37.84501153334721


#Variance
Variance is another number that indicates how spread out the values are. If you take the square root of the variance, you get the standard deviation. Or, if you multiply the standard deviation by itself, you get the variance.

`√1432.25 = 37.85`

Use the NumPy var() method to find the variance:

In [7]:
import numpy

speed = [32,111,138,28,59,77,97]

x = numpy.var(speed)

print(x)

1432.2448979591834


#Percentiles
Percentiles are used in statistics to give you a number that describes the value that a given percent of the values are lower than.

Use the NumPy percentile() method to find the percentiles:

In [8]:
import numpy

ages = [5,31,43,48,50,41,7,11,15,39,80,82,32,2,8,6,25,36,27,61,31]

x = numpy.percentile(ages, 75)

print(x)

43.0
